Add a separate action for removing old wheels (#95)
scientific-python/upload-nightly-action: README.md

# Nightly upload

This is a GitHub Action that uploads (and helps remove) nightly builds to
the [scientific-python nightly channel][], as recommended in

> [SPEC4 — Using and Creating Nightly Wheels][].

Reviewer: These are two separate actions meant for very different audiences. Please do not combine them in the same areas of the README. They should be separated into different sections.

      uses: scientific-python/upload-nightly-action@82396a2ed4269ba06c6b2988bb4fd568ef3c3d6b # 0.6.1
      with:
        artifacts_path: dist

>         anaconda_nightly_upload_token: ${{secrets.ANACONDA_TOKEN}}

Reviewer: Why is this getting changed?

Author: This is to suggest that since projects will use the same API key for their uploads and removals, having "UPLOAD" in the name (`UPLOAD_TOKEN`) wouldn't make sense for removals. Is there a better way to convey this information?

Author: Oh, your comment above of separating the actions into different sections should fix this, let me revert this change.

        anaconda_nightly_upload_token: ${{secrets.UPLOAD_TOKEN}}
```

> [!IMPORTANT]
> Note that we recommend pinning the action against a specific SHA
(rather than a tag), to guard against the unlikely event of upstream
being compromised.

> # Removing old nightly builds

Reviewer 2: ```suggestion
## Removing old nightly builds
```

Reviewer 2: Presuming `#` is kept fot title?

Author: Oh, I was keeping this because this repository would support two major actions. I'll keep just one title for the README.

Author: Addressed in 556e9a6bca0f2d1ff406804ed59829400c6a9fbe.

This repository also ships with an action to ease removals of older nightly wheels from a channel.

To use this functionality, add the following snippet to your workflow:

```yml
jobs:
  steps:
    ...
    - name: Remove old wheels
      uses: scientific-python/upload-nightly-action/remove-wheels@cantknowhashyet # 0.6.0
      with:
        n_latest_uploads: ${{ env.N_LATEST_UPLOADS }}

>         anaconda_nightly_upload_organization: "your-organization"

Reviewer 2: This is called organization, not channel; are these interchangeable terms? If so, use channel; if not, explain the difference.

Author: I'm not quite sure, since they look like they have been used interchangeably. For example, the README outside of this PR:

https://github.com/scientific-python/upload-nightly-action/blob/920fb594ddba437d939052cabd2a59ceb67df265/README.md?plain=1#L48-L67

mentions how one may upload to a different channel, but it uses "organization" as an input.

        anaconda_nightly_token: ${{secrets.ANACONDA_TOKEN}}
```

Which will remove all but the `n_latest_uploads` latest uploads from the channel. This is useful
to avoid hosting outdated development versions, as well as to clean up space.

> Note that the ``scientific-python-nightly-wheels`` channel, specifically, already removes

Reviewer 2: Should we move this up, so the person doesn't read all of this, make a change to their CI, and only then read the part where it's unnecessary?

Author: Addressed in e403c81ce80957498a90530b41b9d19ac6ab16f2.

Author: Yes, that's reasonable – addressed in e403c81ce80957498a90530b41b9d19ac6ab16f2.

        anaconda_nightly_upload_token: ${{secrets.UPLOAD_TOKEN}}
```

> [!IMPORTANT]
> Note that we recommend pinning the action against a specific SHA
(rather than a tag), to guard against the unlikely event of upstream
being compromised.

# Removing old nightly builds

This repository also ships with an action to ease removals of older nightly wheels from a channel.

To use this functionality, add the following snippet to your workflow:

```yml
jobs:
  steps:
    ...
    - name: Remove old wheels
      uses: scientific-python/upload-nightly-action/remove-wheels@cantknowhashyet # 0.6.0
      with:
        n_latest_uploads: ${{ env.N_LATEST_UPLOADS }}

>         anaconda_nightly_upload_token: ${{secrets.UPLOAD_TOKEN}}

Reviewer 2: Add channel to this example, since default channel doesn't need this action.

```

Which will remove all but the `n_latest_uploads` latest uploads from the channel. This is useful
to avoid hosting outdated development versions, as well as to clean up space.

> The channel to remove wheels from is set to the ``scientific-python-nightly-wheels`` channel
> by default. If you are using this channel, please note that this repository will automatically
> clean up old artifacts for you.

Reviewer 2: ```suggestion
Note that the ``scientific-python-nightly-wheels``, specifically, already removes
old artifacts daily. The `remove-wheels` action is therefore intended for use with other channels.
```

If you do not wish to have this automated cleanup, please open an issue on this repository
to be added to the list of packages exempt from it. The current ones are named in 

> `packages-ignore-from-cleanup.txt`.

Reviewer 2: I'd link to an explicit URL here, to make clear which repo we're talking about.

Please refer to the [artifact cleanup policy][] for more information.

> To remove wheels from a different channel, set the ``anaconda_nightly_upload_organization``

Reviewer 2: Include this in the example instead.

with permissions to _Allow write access to the API site_ and _Allow uploads to Standard Python repositories_,
and add the token as a secret to your GitHub repository.

> ## Using a different channel other than ``scientific-python-nightly-wheels``

Reviewer 2: ```suggestion
## Using a channel other than ``scientific-python-nightly-wheels``
```

Author: Oops, thanks!
